### ISSUE TITLE
New clippy::cargo-toml lint category

### ISSUE THREAD

**Issue by vanillajonathan** [rust-lang/rust-clippy#6431]
### What it does

Lints the `Cargo.toml` manifest file and make sure everything is alright.

### Lints

- Cargo.toml — file is missing.
- Cargo.toml — file is misnamed, maybe "cargo.toml or "CARGO.toml", etc
- package.name — starts with contains "use", "using" or "with"
- package.name — contains characters not allowed
- package.version — does not follow semantic versioning
- package.authors — at least one author required
- package.edition — is unknown
- package.documentation — is http not https
- package.readme — file no found
- package.homepage — is http not https
- package.repository — is http not https
- package.license — is not a SPDX 2.1 license expression
- package.license-file — file not found
- package.keywords — contains invalid letters
- package.keywords — exceeds 5 keywords
- package.categories — exceed 5 categories
- package.categories — the category does not exist
- package.build — file not found
- package._foo_ — unknown field
- chapters are in the wrong order
- fields are in the wrong order
- dependencies are not sorted alphabetically

https://doc.rust-lang.org/cargo/reference/manifest.html

**Comment by giraffate** [comment]
FYI Clippy has a [`cargo_common_metadata`](https://rust-lang.github.io/rust-clippy/master/#cargo_common_metadata).

**Comment by taiki-e** [comment]
IIUC `cargo_common_metadata` is for detecting the lack of fields [recommended to include](https://rust-lang.github.io/api-guidelines/documentation.html#c-metadata), not for detecting whether the specified field is valid. I don't think there is lint that detects invalid fields (at least in `clippy::cargo` group).

I find some of these useful, but some feel subjective, so I would suggest separating this into *"what will be rejected by crates.io"* (e.g., exceeds 5 keywords/categories) and *"crates.io will accept but not recommended"*.

Also, I don't think some things need to be linted:

> * Cargo.toml — file is missing.
> * package.name — contains characters not allowed
> * package.version — does not follow semantic versioning

These are probably detected by cargo (during a normal build).

> * package.edition — is unknown

If an edition is not specified, it will be treated as the 2015 edition. And invalid editions are rejected by cargo.
And note that forcing the edition field will force you to lose compatibility with rust before the 2018 edition stabilizes.

> * chapters are in the wrong order
> * fields are in the wrong order

Does the "correct order" exist? IIRC some projects have coding rules for this, but AFAIK, there are no guidelines for this widely used in the rust ecosystem.

> * package.name — starts with contains "use", "using" or "with"

Why are names start with these strings bad?

> * dependencies are not sorted alphabetically

FYI: A tool to do this already exists: [cargo-sort-ck](https://crates.io/crates/cargo-sort-ck)

**Comment by vanillajonathan** [comment]
> Also, I don't think some things need to be linted:
> 
> > * Cargo.toml — file is missing.
> > * package.name — contains characters not allowed
> > * package.version — does not follow semantic versioning

The package manifest states that `package.name` must use only alphanumeric characters or `-` or `_`, and cannot be empty. ([ref](https://doc.rust-lang.org/cargo/reference/manifest.html#the-name-field))

The `package.version` does not need to follow semantic versioning if `package.publish` is set to `false`, but other way the manifest documentation states that [semantic versioning should be used](https://doc.rust-lang.org/cargo/reference/manifest.html#the-version-field).

> > * chapters are in the wrong order
> > * fields are in the wrong order
> 
> Does the "correct order" exist? IIRC some projects have coding rules for this, but AFAIK, there are no guidelines for this widely used in the rust ecosystem.

The correct order would be as listed in the [Cargo.toml manifest documentation](https://doc.rust-lang.org/cargo/reference/manifest.html).

> 
> > * package.name — starts with contains "use", "using" or "with"
> 
> Why are names start with these strings bad?
> 

My bad, I thought the Rust API Guidelines stated so, but this was not for the `package.name` this was for the package features. ([ref](https://rust-lang.github.io/api-guidelines/naming.html#c-feature))

> > * dependencies are not sorted alphabetically
> 
> FYI: A tool to do this already exists: [cargo-sort-ck](https://crates.io/crates/cargo-sort-ck)

There can still be a lint for this. Not all people know about this tool.

**Comment by taiki-e** [comment]
> The package manifest states that `package.name` must use only alphanumeric characters or `-` or `_`, and cannot be empty. ([ref](https://doc.rust-lang.org/cargo/reference/manifest.html#the-name-field))
> 
> The `package.version` does not need to follow semantic versioning if `package.publish` is set to `false`, but other way the manifest documentation states that [semantic versioning should be used](https://doc.rust-lang.org/cargo/reference/manifest.html#the-version-field).

Yeah, but it is detected by cargo, so it doesn't need to be detected by clippy.

> The correct order would be as listed in the [Cargo.toml manifest documentation](https://doc.rust-lang.org/cargo/reference/manifest.html).
> 

Oh, I thought it was just listing the available sections.

**Comment by vanillajonathan** [comment]
> > The correct order would be as listed in the [Cargo.toml manifest documentation](https://doc.rust-lang.org/cargo/reference/manifest.html).
> 
> Oh, I thought it was just listing the available sections.

Yes, I think the order is rather informal and that the  documentation just lists the available sections, but it _might_ be a good idea idea to enforce a particular order just for the sake of consistency.

**Comment by ebroto** [comment]
As @taiki-e said, I think some of the suggestions could be useful. I think most of this can be done in a single lint and does not require a new category. The general idea of the lint could be compliance with `crates.io`; note that there can be alternative crate registries so I think `crates.io` should be stated in the name of the lint somehow. 

I would rule out anything that is already checked by `cargo`. e.g. we should not check that `package.name` "is a valid Rust identifier and not a keyword", but we should check that it uses "only ASCII characters" (among others).

>it might be a good idea idea to enforce a particular order just for the sake of consistency.

I think this should be done as a separate lint. It's unfortunate that with the current lint categories we can't make a distinction between `pedantic` and `style` when creating a `cargo` lint. IMO this would make sense as `pedantic`.

>dependencies are not sorted alphabetically

Same here, I think this is not enforced by crates.io. It could be a separate lint.

>And note that forcing the edition field will force you to lose compatibility with rust before the 2018 edition stabilizes.

Indeed, I think a missing edition should not be linted.

**Comment by vanillajonathan** [comment]
> > dependencies are not sorted alphabetically
> 
> Same here, I think this is not enforced by crates.io. It could be a separate lint.

It is not enforced by crates.io, it just one of my ideas. I personally just happened to think it would be nice if the dependencies were alphabetically sorted.

**Comment by GrygrFlzr** [comment]
> - package.authors — at least one author required

There is an RFC to either deprecate or make optional the authors field, so it might be a good idea to hold off from that lint for now.

See rust-lang/rfcs#3052

**Comment by nyurik** [comment]
In 2026, I think missing `edition` field should be a warning.  Is there any new thoughts on moving this forward, or at least introducing individual lints?
